src/cpp/session/modules/connections/ConnectionsIndexer.cpp

```diff
@@ -0,0 +1,321 @@
+      {
+         std::map<std::string, std::string> fields = parseConnectionsDcf(*it);
+         add(pkgName, fields);
+      }
+   }
+   
+   bool contains(const std::string& package, const std::string& name)
+   {
+      return connections_.count(constructKey(package, name));
+   }
+
+   const ConnectionsIndexEntry& get(const std::string& package, const std::string& name)
+   {
+      return connections_[constructKey(package, name)];
+   }
+   
+   json::Object toJson() const
+   {
+      json::Object object;
+      
+      BOOST_FOREACH(const std::string& key, connections_ | boost::adaptors::map_keys)
+      {
+         object[key] = connections_.at(key).toJson();
+      }
+      
+      return object;
+   }
+
+   std::size_t size() const { return connections_.size(); }
+   
+private:
+   
+   static std::map<std::string, std::string> parseConnectionsDcf(
+       const std::string& contents)
+   {
+      // read and parse the DCF file
+      std::map<std::string, std::string> fields;
+      std::string errMsg;
+      Error error = text::parseDcfFile(contents, true, &fields, &errMsg);
+      if (error)
+         LOG_ERROR(error);
+
+      return fields;
+   }
+
+   static std::string constructKey(const std::string& package, const std::string& name)
+   {
+      return package + "::" + name;
+   }
+
+   std::map<std::string, ConnectionsIndexEntry> connections_;
+};
+
+boost::shared_ptr<ConnectionsRegistry> s_pCurrentConnectionsRegistry =
+    boost::make_shared<ConnectionsRegistry>();
+
+ConnectionsRegistry& connectionsRegistry()
+{
+   return *s_pCurrentConnectionsRegistry;
+}
+
+void updateConnectionsRegistry(boost::shared_ptr<ConnectionsRegistry> pRegistry)
+{
+   s_pCurrentConnectionsRegistry = pRegistry;
+}
+
+class ConnectionsIndexer : public ppe::Indexer
+{
+   void onIndexingStarted()
+   {
+      pRegistry_ = boost::make_shared<ConnectionsRegistry>();
+   }
+   
+   void onWork(const std::string& pkgName, const FilePath& connectionExtensionPath)
+   {
+      pRegistry_->add(pkgName, connectionExtensionPath);
+   }
+   
+   void onIndexingCompleted()
+   {
+      // finalize by indexing current package
+      if (isDevtoolsLoadAllActive())
+      {
+         FilePath pkgPath = projects::projectContext().buildTargetPath();
+         FilePath extensionPath = pkgPath.childPath("inst/rstudio/connections.dcf");
+         if (extensionPath.exists())
+         {
+            std::string pkgName = projects::projectContext().packageInfo().name();
+            pRegistry_->add(pkgName, extensionPath);
+         }
+      }
+
+      // update the connections registry
+      updateConnectionsRegistry(pRegistry_);
+
+      // handle pending continuations
+      json::Object registryJson = connectionsRegistry().toJson();
+      BOOST_FOREACH(json::JsonRpcFunctionContinuation continuation, continuations_)
+      {
+         json::JsonRpcResponse response;
+         response.setResult(registryJson);
+         continuation(Success(), &response);
+      }
```

Comment:
This would be prudent, although this registry will get cleaned up every time the indexer is re-run so in practice we'll just have a single extra registry copy hanging around.